qiskit_optimization/algorithms/qrao/magic_rounding.py

```
@@ -0,0 +1,469 @@
+            )
+
+        if self._sampler.options.get("shots") is None:
+            raise ValueError(
+                "Magic rounding requires the sampler to be configured with a number of shots."
+            )
+        self._shots = self._sampler.options.shots
+
+        if self.basis_sampling == "uniform":
+            # uniform sampling
+            bases, basis_shots = self._sample_bases_uniform(q2vars, vars_per_qubit)
+        else:
+            # weighted sampling
+            if expectation_values is None:
+                raise ValueError(
+                    "Magic rounding with weighted sampling requires the expectation values of the "
+                    "``RoundingContext`` to be available, but they are not."
+                )
+            bases, basis_shots = self._sample_bases_weighted(
+                q2vars, expectation_values, vars_per_qubit
+            )
+
+        # For each of the Magic Bases sampled above, measure
+        # the appropriate number of times (given by basis_shots)
+        # and return the circuit results
+        basis_counts = self._evaluate_magic_bases(circuit, bases, basis_shots, vars_per_qubit)
+        # keys will be configurations of decision variables
+        # values will be total number of observations.
+        soln_counts = self._compute_dv_counts(basis_counts, bases, var2op, vars_per_qubit)
+
+        soln_samples = [
+            SolutionSample(
+                x=np.asarray([int(bit) for bit in soln]),
+                fval=rounding_context.encoding.problem.objective.evaluate(
+                    [int(bit) for bit in soln]
+                ),
+                probability=count / self._shots,
+                status=OptimizationResultStatus.SUCCESS
+                if rounding_context.encoding.problem.is_feasible([int(bit) for bit in soln])
+                else OptimizationResultStatus.INFEASIBLE,
+            )
+            for soln, count in soln_counts.items()
+        ]
+
+        assert np.isclose(
```

You should not use assert. If necessary, could you raise an error?